### ISSUE TITLE
Initialization of generic classes with type parameter default values is broken

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ty#1156]
### Summary

I identified another problem that "blocks" the merge of https://github.com/astral-sh/ruff/pull/18007 (in the sense that there are hundreds of false positives caused by this). It can be reproduced on `main` by explicitly annotating `self: Self`, and occurs when trying to initialize a generic class with a type parameter default value:

```py
from typing import Self

class MyClass[T = int]:
    def __init__(self: Self, x: T) -> None:
        self.x = x

# error: Argument to bound method `__init__` is incorrect: Argument type `MyClass[T@MyClass]` does not satisfy upper bound of type variable `Self`
MyClass(1)
```
https://play.ty.dev/adc72fc1-667d-4956-aa06-ed72e1403e1f

### Version

Current main (25853e237)

**Comment by dcreager** [comment]
It looks like we're using the typevar default instead of the type inferred by the arguments:

```py
from typing import Self

class MyClass[T = str]:
    def __init__(self: Self, x: T) -> None:
        self.x = x

reveal_type(MyClass(1))  # revealed: MyClass[str]
```

https://play.ty.dev/1f512f4b-b072-4844-b2dc-314b9b1b2598
